cosmoz-default-tree.html

```diff
@@ -14,143 +27,249 @@
+		var results = [];
 
-		if (propertyName === 'pathLocator') {
-			return this._getPathByPathLocator(propertyValue);
-		}
+		// Defaults
+		nodes = nodes || this._roots;
 
-		var node = this._searchAllRoots(propertyName, propertyValue);
-		if (node !== undefined) {
-			return this._getPathByPathLocator(node.pathLocator);
-		}
+		nodes.some(function (node) {
+			results = results.concat(this.search(node, propertyValue, options));
+			return options.firstHitOnly && results.length > 0;
+		}, this);
+
+		return results;
 	};
 
-	Cosmoz.DefaultTree.prototype.getChildren = function (node) {
-		if (node === undefined) {
-			return;
+	/**
+	 * Returns the node of a given path.
+	 * @return {object}
+	 * @param {string} pathLocator (The string which describes the path. e.g. "1.2.9")
+	 * @param {object} nodeObj [this._treeData] (The object the search should be based on.)
+	 * @param {string} pathLocatorSeparator [this.pathLocatorSeparator] (The string which separates the path. e.g ".")
+	*/
+	Cosmoz.DefaultTree.prototype.getNodeByPathLocator = function (pathLocator, nodeObj, pathLocatorSeparator) {
+		if (!pathLocator) {
+			return this._roots;
 		}
-		return _objectValues(node.children);
+
+		// Defaults
+		nodeObj = nodeObj || this._treeData;
+		pathLocatorSeparator = pathLocatorSeparator || this.pathLocatorSeparator;
+
+		var	pathNodes = this.getPathNodes(pathLocator, nodeObj, pathLocatorSeparator);
+		return pathNodes && pathNodes.pop();
 	};
 
-	Cosmoz.DefaultTree.prototype.getProperty = function (node, propertyName) {
-		if (node === undefined || propertyName === undefined) {
-			return;
+	/**
+	 * Returns the nodes on a given path.
+	 * A valid path 1.2.3 should return the items [1, 2, 3]
+	 * - path 1.2.3.3 should return [1, 2, 3, undefined]
+	 * - path 0.1.2.3 should return [1, 2, 3]
+	 * - path 0.1.5.3 should return [1, undefined, undefined]
+	 * @return {array}
+	 * @param {string} pathLocator (The string which describes the path. e.g. "1.2.9")
+	 * @param {object} nodeObj [this._treeData] (The object the search should be based on.)
+	 * @param {string} pathLocatorSeparator [this.pathLocatorSeparator] (The string which separates the path.)
+	*/
+	Cosmoz.DefaultTree.prototype.getPathNodes = function (pathLocator, nodeObj, pathLocatorSeparator) {
+		if (!pathLocator) {
+			return this._roots;
 		}
 
-		return node[propertyName];
-	};
+		// Defaults
+		pathLocatorSeparator = pathLocatorSeparator || this.pathLocatorSeparator;
 
-	Cosmoz.DefaultTree.prototype._searchAllRoots = function (propertyName, propertyValue) {
-		var i,
+		var path = pathLocator.split(pathLocatorSeparator),
+			pathSegment = nodeObj || this._treeData,
+			nodes = [],
 			node;
 
-		for (i = 0; i < this._roots.length; i+=1) {
-			node = this.search(this._roots[i], propertyName, propertyValue);
-			if (node !== undefined) {
-				return node;
+		// Get the nodes on the path
+		path.forEach(function (nodeKey) {
```

Comment:
Well I think you can stop searching any more roots if you find a result _without_ any `undefined` parts, otherwise we have to figure out what path has least number of undefined parts.. whatever that will help.. :)